Best practices/React/Organization.md

```diff
@@ -4,11 +4,109 @@
 
 This section describes a sensible starting point for organizing a React application that supports server-side rendering.
 
-## `/app`
+## Application
 
-### `/app/components` and `<Component>/components`
+### React Components and Sub-Components
+
+React Components should always exist in a `components/` folder, but the component's internal structure may change depending on their complexity.
+
+
+#### 1. Start with a file and a test
+
+At its simplest a component will exist as a component and a test at the `components/` level.
+
+If you foresee this component growing quickly in complexity or needing a subcomponent, feel free to jump straight to #2.
+```
+.
+└── components
+    ├── ...
+    ├── MyComponent.js
+    └── tests
+        ├── ...
+        └── MyComponent.test.js
+```
+
+
+#### 2. Isolate your component to a folder (of the same name) as the component grows
+
+```
+├── component.js
+└── components
+    └── MyComponent
+        ├── MyComponent.js
+        ├── MyComponent.scss
+        └── MyComponent.test.js
+```
+
+When you move your component into it's own folder, you will need to figure out how to maintain all existing import paths.
+
+```
+import {MyComponent} from './components/MyComponent';
+// would need to become 👇
+import {MyComponent} from './components/MyComponent/MyComponent';
+```
+
+You have several options to fix this problem:
+
+1. rename the component's filename from `MyComponent.js` to `index.js` to
+2. create an `index.js` inside the component's folder that exports `MyComponent.js`
+3. create a `components.js` file next to the component's folder that exports each component in the `components/` folder.
+
+#### 3. As your component grows, group like-files in folders.
+
+Common types to group inside a component:
```

Comment:
Can we represent `hooks` here as well by way of a `hooks` directory or `hooks.ts` file. You can link to the section below for more details.

Reply:
Great idea 👍